src/xenia/kernel/xfile.cc

```diff
@@ -160,18 +160,76 @@ X_STATUS XFile::Read(uint32_t buffer_guest_address, uint32_t buffer_length,
     }
   }
 
+  if (out_bytes_read) {
+    *out_bytes_read = uint32_t(bytes_read);
+  }
+
+  if (notify_completion) {
+    XIOCompletion::IONotification notify;
+    notify.apc_context = apc_context;
+    notify.num_bytes = uint32_t(bytes_read);
+    notify.status = result;
+
+    NotifyIOCompletionPorts(notify);
+
+    async_event_->Set();
+  }
+
+  return result;
+}
+
+X_STATUS XFile::ReadScatter(uint32_t buffers_guest_address, uint32_t length,
+                            uint64_t byte_offset, uint32_t* out_bytes_read,
+                            uint32_t apc_context) {
+  X_STATUS result = X_STATUS_SUCCESS;
+
+  // buffers points to an array of buffer pointers of type
+  // "FILE_SEGMENT_ELEMENT", but they can just be treated as normal pointers
+  xe::be<uint32_t>* buffers = reinterpret_cast<xe::be<uint32_t>*>(
+      memory()->TranslateVirtual(buffers_guest_address));
+
+  // TODO: not sure if this is meant to change depending on buffer address?
+  uint32_t page_size = 4096;
+
+  uint32_t read_total = 0;
+  uint32_t read_remain = length;
+  while (read_remain) {
```

Comment:
Can ReadFileScatter be used on Windows instead of the loop?

Reply:
Looks like ReadFileScatter only works asynchronously, but AFAIK all file IO is done synchronous atm without the FILE_FLAG_OVERLAPPED flag that ReadFileScatter needs.. If async support is ever added it might be worth looking into though, I'll add a note to the unused async part of NtReadFileScatter about it.